### PR TITLE
Add support for updating POS app through Greenfield API

### PR DESCRIPTION
**Pull request by bolatovumar**
With this PR you will be able to update a POS app through Greenfield API.

Part of #3458

### CODE REVIEW

File: BTCPayServer.Client/Models/CreateAppRequest.cs [hunk 1]
```diff
@@ -25,9 +25,9 @@ public class CreatePointOfSaleAppRequest : CreateAppRequest
         public string Template { get; set; } = null;
         [JsonConverter(typeof(StringEnumConverter))]
         public PosViewType DefaultView { get; set; }
-        public bool ShowCustomAmount { get; set; } = true;
-        public bool ShowDiscount { get; set; } = true;
-        public bool EnableTips { get; set; } = true;
+        public bool? ShowCustomAmount { get; set; } = null;
```
Comment (bolatovumar) [inline BTCPayServer.Client/Models/CreateAppRequest.cs:28]: Fallback in the code instead of here and mark these as optional as they are.

File: BTCPayServer.Tests/GreenfieldAPITests.cs [hunk 2]
```diff
@@ -89,7 +89,7 @@ public async Task ApiKeysControllerTests()
             await user.MakeAdmin();
             var client = await user.CreateClient(Policies.CanViewProfile);
             var clientBasic = await user.CreateClient();
-            //Get current api key
```
Comment (bolatovumar) [inline BTCPayServer.Tests/GreenfieldAPITests.cs:92]: The whitespaces were auto-trimmed. Ignore all of these.

File: BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs [hunk 3]
```diff
@@ -22,33 +23,38 @@ public class GreenfieldAppsController : ControllerBase
...
             _storeRepository = storeRepository;
+            _currencies = currencies;
         }
 
         [HttpPost("~/api/v1/stores/{storeId}/apps/pos")]
         [Authorize(Policy = Policies.CanModifyStoreSettings, AuthenticationSchemes = AuthenticationSchemes.Greenfield)]
         public async Task<IActionResult> CreatePointOfSaleApp(string storeId, CreatePointOfSaleAppRequest request)
         {
-            var validationResult = Validate(request);
+            var store = await _storeRepository.FindStore(storeId);
```
Comment (bolatovumar) [inline BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs:45]: Moved the store validation to the top since it doesn't make sense to first validate the request and only then validate that the store we are trying to create an app for exists, should be the other way around.

File: BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs [hunk 4]
```diff
@@ -60,11 +66,11 @@ public async Task<IActionResult> CreatePointOfSaleApp(string storeId, CreatePoin
...
-                ShowCustomAmount = request.ShowCustomAmount,
-                ShowDiscount = request.ShowDiscount,
-                EnableTips = request.EnableTips,
-                Currency = request.Currency ?? defaultCurrency,
-                Template = request.Template,
+                ShowCustomAmount = request.ShowCustomAmount ?? true,
+                ShowDiscount = request.ShowDiscount ?? true,
+                EnableTips = request.EnableTips ?? true,
+                Currency = request.Currency,
+                Template = request.Template != null ? _appService.SerializeTemplate(_appService.Parse(request.Template, request.Currency)) : null,
```
Comment (bolatovumar) [inline BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs:73]: Previously the template was saved raw which is wrong cuz you can end up saving an invalid template. I added validation for this as well now.

File: BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs [hunk 5]
```diff
@@ -74,18 +80,74 @@ public async Task<IActionResult> CreatePointOfSaleApp(string storeId, CreatePoin
...
+            var validationResult = ValidatePOSAppRequest(request);
+            if (validationResult != null)
+            {
+                return validationResult;
+            }
+
+            app.Name = request.AppName ?? app.Name;
+            app.SetSettings(new PointOfSaleSettings
+            {
+                Title = request.Title ?? settings.Title,
```
Comment (bolatovumar) [inline BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs:115]: Use the request properties or fallback to what exists on the app already if not provided.

File: BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs [hunk 6]
```diff
@@ -130,7 +192,35 @@ private PointOfSaleAppData ToModel(AppData appData)
             };
         }
 
-        private IActionResult? Validate(CreateAppRequest request)
+        private IActionResult? ValidatePOSAppRequest(CreatePointOfSaleAppRequest request)
+        {
+            var validationResult = ValidateCreateAppRequest(request);
+            if (request.Currency != null && _currencies.GetCurrencyData(request.Currency, false) == null)
```
Comment (bolatovumar) [inline BTCPayServer/Controllers/GreenField/GreenfieldAppsController.cs:198]: Added previously missing validation for currency and template validity.

**Review by Kukks — CHANGES_REQUESTED** [review]
Please do a full UPDATE (Http PUT not PATCH) instead of selective property updating, as we have not defined this update mechanism for any other entity or endpoint.

**Comment by NicolasDorier** [comment]
I think that it is fine. While it is strange to have our first PATCH, it doesn't mean that is bad.

**Comment by NicolasDorier** [comment]
@bolatovumar I think this PATCH doesn't work as expected. Because you call `ValidateCreateAppRequest` which doesn't expect to stumble across null values.

**Comment by NicolasDorier** [comment]
@bolatovumar I thought with it with @Kukks , I think that it should be PUT.

Not only because the PATCH has already some bug here, but because I believe PATCH for all routes with a GET and PUT can be created automatically the following way:

1. Do the json from the GET
2. Patch the json with the JObject of the PATCH
3. Do the PUT.

It can be generalized for every routes having both PUT and GET, without corner case to handle manually.

**Comment by bolatovumar** [comment]
@Kukks @NicolasDorier alright, updated to be a `PUT` instead. Failing test is unrelated.

Capture

File: BTCPayServer/Controllers/GreenField/LocalBTCPayServerClient.cs [hunk 7]
```diff
@@ -1103,6 +1103,14 @@ await GetController<GreenfieldPullPaymentController>()
                 await GetController<GreenfieldAppsController>().CreatePointOfSaleApp(storeId, request));
         }
 
+        public override async Task<PointOfSaleAppData> PutPointOfSaleApp(
```
[the remark this answers is not in the document]
Comment (bolatovumar) [reply BTCPayServer/Controllers/GreenField/LocalBTCPayServerClient.cs]: Updated in https://github.com/btcpayserver/btcpayserver/pull/3979/commits/0dfbc3725f47cae92ed9d1032ed30b0b60c6c1f3

File: BTCPayServer/wwwroot/swagger/v1/swagger.template.apps.json [hunk 8]
```diff
@@ -17,128 +17,83 @@
...
+                    "name": "appId",
+                    "in": "path",
+                    "required": true,
+                    "description": "App ID",
+                    "schema": {
+                        "type": "string"
+                    }
+                }
+            ],
+            "patch": {
```
[the remark this answers is not in the document]
Comment (bolatovumar) [reply BTCPayServer/wwwroot/swagger/v1/swagger.template.apps.json]: Updated in https://github.com/btcpayserver/btcpayserver/pull/3979/commits/a8d6f78955569f1e2f2352f0f24c06c60e0ece6a

**Review by Kukks — APPROVED** [review]

**Comment by bolatovumar** [comment]
Failing test is still unrelated.

**Comment by bolatovumar** [comment]
@NicolasDorier ping. I rebased the branch and now tests are green.

**Review by Kukks — APPROVED** [review]
